daemon/daemon.go

```diff
@@ -1086,23 +1084,56 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
 
 	d := h.daemon
 
-	if err := d.conf.Opts.Validate(params.Configuration); err != nil {
+	if err := d.conf.Opts.Validate(params.Configuration.Mutable); err != nil {
 		return apierror.Error(PatchConfigBadRequestCode, err)
 	}
 
-	changes := d.conf.Opts.Apply(params.Configuration, changedOption, d)
-	log.Debugf("Applied %d changes", changes)
-
-	// Check explicitly for endpoint.OptionPolicy updates because its state
-	// is coupled with config's EnablePolicy flag.
-	_, ok := params.Configuration[endpoint.OptionPolicy]
-	if ok {
-		if config.Opts.IsEnabled(endpoint.OptionPolicy) && config.EnablePolicy != endpoint.AlwaysEnforce {
-			config.EnablePolicy = endpoint.AlwaysEnforce
-		} else if !config.Opts.IsEnabled(endpoint.OptionPolicy) && config.EnablePolicy != endpoint.NeverEnforce {
-			config.EnablePolicy = endpoint.NeverEnforce
+	// Only configure PolicyEnforcement through PolicyEnforcement flag, not Policy flag.
+	if _, ok := params.Configuration.Mutable[endpoint.OptionPolicy]; ok {
+		msg := fmt.Errorf("Please configuration policy enforcement for the daemon using \"PolicyEnforcement\" instead of \"Policy\"")
+		log.Warningf("%s", msg)
+		return apierror.Error(PatchConfigBadRequestCode, msg)
+	}
+
+	var policyEnforcementChanged bool
+
+	// Update policy enforcement configuration if needed.
+	config.EnablePolicyMU.RLock()
+	oldEnforcementValue := config.EnablePolicy
+	config.EnablePolicyMU.RUnlock()
+	enforcement := params.Configuration.PolicyEnforcement
+
+	// Only update if value provided for PolicyEnforcement.
+	if enforcement != "" {
+		switch enforcement {
+		case endpoint.NeverEnforce, endpoint.DefaultEnforcement, endpoint.AlwaysEnforce:
+			// If the policy enforcement configuration has indeed changed, we have
+			// to regenerate endpoints and update daemon's configuration.
+			if enforcement != oldEnforcementValue {
+				// Only allow configuration updates for PolicyEnforcement to occur again once configuration has been applied
+				config.EnablePolicyMU.Lock()
+				defer config.EnablePolicyMU.Unlock()
+				config.EnablePolicy = enforcement
+
+				d.GetPolicyRepository().Mutex.RLock()
+				_, policyEnforcementChanged = d.EnablePolicyEnforcement()
+				d.GetPolicyRepository().Mutex.RUnlock()
+				d.TriggerPolicyUpdates([]policy.NumericIdentity{})
+			}
```

Comment:
I fixed the race condition, but not deferring the unlocking of `EnablePolicyMU` introduced very weird behavior on my end, including putting endpoints into 'not-ready' state due to compilation errors when trying to regenerate the endpoints. I can reproduce this fairly easily by changing the locking logic. I checked the generated header files / C code when I saw that the endpoints couldn't regenerate and I couldn't figure out why the compiler was complaining.